panic: Unable to flush sub store: pq: duplicate key value violates unique constraint "pk_msgpendingkey"

Got this error yesterday and today, and it's fixed after deleting the table contents by myself.

```
panic: Unable to flush sub store: pq: duplicate key value violates unique constraint "pk_msgpendingkey"
goroutine 22 [running]:
github.com/nats-io/nats-streaming-server/server.(*StanServer).ioLoop(0xc0001e6000, 0xc000146630)
    /home/admin/nats/src/github.com/nats-io/nats-streaming-server/server/server.go:3782 +0xb4d
created by github.com/nats-io/nats-streaming-server/server.(*StanServer).startIOLoop
    /home/admin/nats/src/github.com/nats-io/nats-streaming-server/server/server.go:3611 +0xcc
```

Version :
version | 1.3.0
proto | 1
go | go1.11.4
Postgres : 10 on AWS RDS

--- a/server/server_delivery_test.go
+++ b/server/server_delivery_test.go
@@ -19,10 +19,12 @@ import (
 	"testing"
 	"time"
 
+	natsdTest "github.com/nats-io/gnatsd/test"
 	"github.com/nats-io/go-nats"
 	"github.com/nats-io/go-nats-streaming"
 	"github.com/nats-io/go-nats-streaming/pb"
 	"github.com/nats-io/nats-streaming-server/stores"
+	"github.com/nats-io/nats-streaming-server/test"
 )
 
 func testStalledDelivery(t *testing.T, typeSub string) {
@@ -252,4 +254,70 @@ func TestDeliveryWithGapsInSequence(t *testing.T) {
 		t.Fatalf(e.Error())
 	default:
 	}
+}
+
+func TestPersistentStoreSQLSubsPendingRows(t *testing.T) {
+	source := testSQLSource
+	if persistentStoreType != stores.TypeSQL {
+		// If not running tests with `-persistent_store sql`,
+		// initialize few things and default to MySQL.
+		source = testDefaultMySQLSource
+		sourceAdmin := testDefaultMySQLSourceAdmin
+		if err := test.CreateSQLDatabase(testSQLDriver, sourceAdmin,
+			source, testSQLDatabaseName); err != nil {
+			t.Fatalf("Error setting up test for SQL: %v", err)
+		}
+		defer test.DeleteSQLDatabase(testSQLDriver, sourceAdmin, testSQLDatabaseName)
+	}
+
+	cleanupDatastore(t)
+	defer cleanupDatastore(t)
+
+	ns := natsdTest.RunDefaultServer()
+	defer ns.Shutdown()
+
+	opts := GetDefaultOptions()
+	opts.NATSServerURL = "nats://localhost:4222"
+	opts.StoreType = stores.TypeSQL
+	opts.SQLStoreOpts.Driver = testSQLDriver
+	opts.SQLStoreOpts.Source = source
+	s := runServerWithOpts(t, opts, nil)
+	defer shutdownRestartedServerOnTestExit(&s)
+
+	sc := NewDefaultConnection(t)
+	defer sc.Close()
+
+	// Create a regular sub and a durable.
+	if _, err := sc.Subscribe("foo", func(_ *stan.Msg) {},
+		stan.SetManualAckMode(),
+		stan.MaxInflight(5000)); err != nil {
+		t.Fatalf("Error on subscribe: %v", err)
+	}
+	ch := make(chan bool, 1)
+	dur, err := sc.Subscribe("foo",
+		func(_ *stan.Msg) {
+			ch <- true
+		},
+		stan.SetManualAckMode(),
+		stan.DurableName("dur"))
+	if err != nil {
+		t.Fatalf("Error on subscribe: %v", err)
+	}
+	// Publish a message
+	sc.Publish("foo", []byte("hello"))
+	if err := Wait(ch); err != nil {
+		t.Fatalf("Did not get our message")
+	}
+	dur.Close()
+	// Produce another message
+	sc.Publish("foo", []byte("hello"))
+
+	// Restart the server
+	s.Shutdown()
+	s = runServerWithOpts(t, opts, nil)
+	// Bombard the running subscriber with messages.
+	for i := 0; i < 3000; i++ {
+		sc.PublishAsync("foo", []byte("hello"), nil)
+	}
+	waitForAcks(t, s, clientName, 1, 3002)
 }

--- a/stores/sqlstore.go
+++ b/stores/sqlstore.go
@@ -104,10 +104,10 @@ var sqlStmts = []string{
 	"INSERT INTO Channels (id, name, maxmsgs, maxbytes, maxage) VALUES (?, ?, ?, ?, ?)",                          // sqlAddChannel
 	"INSERT INTO Messages VALUES (?, ?, ?, ?, ?)",                                                                // sqlStoreMsg
 	"SELECT timestamp, data FROM Messages WHERE id=? AND seq=?",                                                  // sqlLookupMsg
-	"SELECT seq FROM Messages WHERE id=? AND timestamp>=? LIMIT 1",                                               // sqlGetSequenceFromTimestamp
+	"SELECT seq FROM Messages WHERE id=? AND timestamp>=? ORDER BY seq LIMIT 1",                                  // sqlGetSequenceFromTimestamp
 	"UPDATE Channels SET maxseq=? WHERE id=?",                                                                    // sqlUpdateChannelMaxSeq
 	"SELECT COUNT(seq), COALESCE(MAX(seq), 0), COALESCE(SUM(size), 0) FROM Messages WHERE id=? AND timestamp<=?", // sqlGetExpiredMessages
-	"SELECT timestamp FROM Messages WHERE id=? AND seq>=? LIMIT 1",                                               // sqlGetFirstMsgTimestamp
+	"SELECT timestamp FROM Messages WHERE id=? AND seq>=? ORDER BY seq LIMIT 1",                                  // sqlGetFirstMsgTimestamp
 	"DELETE FROM Messages WHERE id=? AND seq<=?",                                                                 // sqlDeletedMsgsWithSeqLowerThan
 	"SELECT size FROM Messages WHERE id=? AND seq=?",                                                             // sqlGetSizeOfMessage
 	"DELETE FROM Messages WHERE id=? AND seq=?",                                                                  // sqlDeleteMessage
@@ -1947,11 +1947,14 @@ func (ss *SQLSubStore) deleteSubPendingRow(subid, rowid uint64) error {
 func (ss *SQLSubStore) recoverPendingRow(rows *sql.Rows, sub *spb.SubState, ap *sqlSubAcksPending, pendingAcks PendingAcks,
 	gcedRows map[uint64]struct{}) error {
 	var (
-		seq, lastSent           uint64
+		rowID, seq, lastSent    uint64
 		pendingBytes, acksBytes []byte
 	)
-	if err := rows.Scan(&ss.curRow, &seq, &lastSent, &pendingBytes, &acksBytes); err != nil && err != sql.ErrNoRows {
+	if err := rows.Scan(&rowID, &seq, &lastSent, &pendingBytes, &acksBytes); err != nil && err != sql.ErrNoRows {
 		return err
+	}
+	if rowID > ss.curRow {
+		ss.curRow = rowID
 	}
 	// If seq is non zero, this was created from a non-buffered run.
 	if seq > 0 {
@@ -1963,7 +1966,7 @@ func (ss *SQLSubStore) recoverPendingRow(rows *sql.Rows, sub *spb.SubState, ap *
 		var row *sqlSubsPendingRow
 		if ap != nil {
 			row = &sqlSubsPendingRow{
-				ID:   ss.curRow,
+				ID:   rowID,
 				msgs: sqlSeqMapPool.Get().(map[uint64]struct{}),
 			}
 			ap.lastSent = lastSent

--- a/stores/sqlstore_test.go
+++ b/stores/sqlstore_test.go
@@ -1239,6 +1239,73 @@ func TestSQLSubStoreCachingAndRecovery(t *testing.T) {
 	testSQLCheckPendingRow(t, db, subID)
 }
 
+func TestSQLSubStoreSubsPendingNoDuplicateForRowID(t *testing.T) {
+	if !doSQL {
+		t.SkipNow()
+	}
+	sqlMaxPendingAcks = 5
+	// "disable" timer based flusing for now
+	sqlSubStoreFlushInterval = time.Hour
+	defer func() {
+		sqlMaxPendingAcks = sqlDefaultMaxPendingAcks
+		sqlSubStoreFlushInterval = sqlDefaultSubStoreFlushInterval
+	}()
+
+	cleanupSQLDatastore(t)
+	defer cleanupSQLDatastore(t)
+
+	// Tests are by default not using caching, so this
+	// test has to be explicit.
+	s, err := NewSQLStore(testLogger, testSQLDriver, testSQLSource, nil, SQLNoCaching(false))
+	if err != nil {
+		t.Fatalf("Error creating store: %v", err)
+	}
+	defer s.Close()
+	info := testDefaultServerInfo
+	if err := s.Init(&info); err != nil {
+		s.Close()
+		t.Fatalf("Error on Init: %v", err)
+	}
+
+	cname := "foo"
+
+	cs := storeCreateChannel(t, s, cname)
+	for i := uint64(1); i < 10; i++ {
+		storeMsg(t, cs, cname, i, []byte("hello"))
+	}
+
+	subID1 := storeSub(t, cs, cname)
+	subID2 := storeSub(t, cs, cname)
+
+	storeSubPending(t, cs, cname, subID1, 1)
+	storeSubFlush(t, cs, cname)
+
+	storeSubPending(t, cs, cname, subID2, 1)
+	storeSubFlush(t, cs, cname)
+
+	storeSubPending(t, cs, cname, subID1, 2)
+	storeSubFlush(t, cs, cname)
+
+	s.Close()
+	s, err = NewSQLStore(testLogger, testSQLDriver, testSQLSource, nil, SQLNoCaching(false))
+	if err != nil {
+		t.Fatalf("Error creating store: %v", err)
+	}
+	defer s.Close()
+	state, err := s.Recover()
+	if err != nil {
+		t.Fatalf("Error recovering state: %v", err)
+	}
+	cs = getRecoveredChannel(t, state, cname)
+
+	// We have sub1 that has row 1 and 3, and sub2 that has row 2.
+	// We now cause sub1 to add a new row since we had up to the max.
+	// We want to make sure that we don't use "3" for the new row id
+	// but use a new one (otherwise there would be a duplicate entry
+	// "1-3" for primary key in subspending table).
+	storeSubPending(t, cs, cname, subID1, 3, 4, 5, 6, 7)
+}
+
 func TestSQLMaxConnections(t *testing.T) {
 	if !doSQL {
 		t.SkipNow()

